### PR TITLE
mgmt: hawkbit: autohandler and shell optimize

### PR DESCRIPTION
**Pull request by maass-hamburg**
also use the workqueue for the run of hawkBit, when requested via the shell. This way the run via the shell will use the same thread as the autohandler run. We also can eliminate redundant code. 

This has mainly the benefit of not having to extend the stack size of the shell threads. Explicitly when having multiple shell instances, then the stack of every instance has to be bigger. 

This is needed for https://github.com/zephyrproject-rtos/zephyr/pull/70787 as it moves the buffers from global to the stack.

### CODE REVIEW

File: subsys/mgmt/hawkbit/shell.c
Comment (ycsin) [inline subsys/mgmt/hawkbit/shell.c:21]: This is not equal to what we had before - `shell_fprintf` would 'reply' to the shell instance where the command came from, `LOG_*` would be sending to whatever log backends configured with the target verbosity.

So if only LOG UART backend is enabled and command is coming from shell_telnet, the user will not be able to see the messages.

**Review by ycsin — CHANGES_REQUESTED** [review]

File: subsys/mgmt/hawkbit/shell.c
Comment (ycsin) [inline subsys/mgmt/hawkbit/shell.c:21]: Are you going to see the same message twice? One from the log and the other one from the shell?

File: subsys/mgmt/hawkbit/hawkbit.c
Comment (maass-hamburg) [reply subsys/mgmt/hawkbit/hawkbit.c:97]: so that it isn't initialised in `hawkbit_autohandler()` and `hawkbit_autohandler()` can be executed more that one without reinitialising again. For maybe converting hawkBit to be instance-based, which will need lot of changes, initialising it statically is not the problem, it would be this being a global variable and that's not changing with this PR. Also making hawkbit instance-based would not be the optimal solution, as hawkBit already supports having multiple software modules as part of a distribution, where one module is the OS and another can be the for settings. That would also be much easier to implement.

File: subsys/mgmt/hawkbit/shell.c
Comment (maass-hamburg) [reply subsys/mgmt/hawkbit/shell.c:21]: yes, if you have the log output activated for that shell. But if you have other log backends activated, like syslog it should also get the logs, independent of the reason why hawkbit was executed.

Comment (ycsin) [reply subsys/mgmt/hawkbit/shell.c:22]: Is there a better way to refactor this mechanism so that it only prints the result to the shell instance that triggered it?

Comment (maass-hamburg) [reply subsys/mgmt/hawkbit/shell.c:25]: IMO it is good, that the autohandler, sending these messages to all log backends, independent of the reason why it was executed (shell or scheduled). It's more like a feature. 
I don't think it is easily possible to omit the logs from the autohandler just for the shell in which it was started, while still sending the log to all other log backends, including the other shells, when they are enabled as log backend.

File: include/zephyr/mgmt/hawkbit.h
Comment (kartben) [inline include/zephyr/mgmt/hawkbit.h:135]: use `@retval`

File: include/zephyr/mgmt/hawkbit.h
Comment (kartben) [inline include/zephyr/mgmt/hawkbit.h:110]: make this one first in the list?

**Review by kartben — DISMISSED** [review]

File: include/zephyr/mgmt/hawkbit.h
Comment (maass-hamburg) [reply include/zephyr/mgmt/hawkbit.h:113]: `retval` would be the better one here. https://www.doxygen.nl/manual/commands.html#cmdretval
`retval` and `return` are not the same.

Comment (kartben) [reply include/zephyr/mgmt/hawkbit.h]: What @maass-hamburg said :)
See the difference between the two below: 

image

image

File: subsys/mgmt/hawkbit/hawkbit.c
Comment (ycsin) [inline subsys/mgmt/hawkbit/hawkbit.c:1504]: Do you expect anyone would use anything other than `__UINT32_MAX__` for `events`?

Comment (maass-hamburg) [reply subsys/mgmt/hawkbit/hawkbit.c:1504]: I can think of some people only using `BIT(HAWKBIT_NO_UPDATE)`

**Review by ycsin — CHANGES_REQUESTED** [review]

**Comment by maass-hamburg** [comment]
Rebased to resolve conflicts.

File: doc/releases/release-notes-3.7.rst
Comment (aescolar) [inline doc/releases/release-notes-3.7.rst:310]: Nit
```suggestion
      function will block until the autohandler is finished. The ``event`` parameter can be set to
      ``UINT32_MAX`` to wait for the autohandler to finish one run, or :c:macro:`BIT`
```
Anyhow, the whole "This  function will block until the autohandler is finished" seems a bit redundant.

**Review by aescolar — COMMENTED** [review]
ok'ish for the release notes and migration guide.
Just a minor suggestion for improvement.
Though, if it were me, I would probably be less verbose and just list which new APIs have been added and rely on the link to their doxygen generated docs, instead of reexplaining it in the release notes.

File: doc/releases/release-notes-3.7.rst
Comment (aescolar) [inline doc/releases/release-notes-3.7.rst:311]:
```suggestion
      together with a value from :c:enum:`hawkbit_response` to wait for a specific event.
```

**Comment by maass-hamburg** [comment]
ping @ycsin

**Review by ycsin — COMMENTED** [review]
I saw that you made some changes to the autohandler stuff, will take a look at those later

Meanwhile here are some quick comments.

The commits should try to explain why you did something in a summary if possible, msg like `"seperate hawkbit.h into different headers."` describes what you did (which we can see from the commit itself), so it didn't provide extra values to people wondering why is this change required

**Review by ycsin — DISMISSED** [review]
I like the fact that we do not need a big stack size for the shell in this PR, but this comes at the expense of multiple additional public functions to meddle with how the hawkBit schedule itself, I'd much rather steer clear of potential racing issues that may arise in the future.

IMHO, while this potentially provides more features to the users, my impression with the built-in autohandler & the shell is that they are there for users to get started with something simple. What I've done in the past is that I'll create my own hawkbit task to do exactly what I need.

My concern with these new features is that they look very 'application' to me (the hawkBit sample basically duplicated the autohandler), I'd prefer hawkBit to stay as a platform/framework/library and let the users decide what's best for them.

From that perspective, these features that we add upstream could restrict the way user uses the hawkBit subsystem - I don't think we will ever satisfy everyone and therefore my expectation is that this trend will likely continue - adding more public functions, each does a very specific thing, attempting to tie up the loose ends, which we will have to maintain.

I understand that this PR is significant as it unblocks #70787 (I didn't really look at that in detail yet, but SMF sounds interesting to me). But if the autohandler or the shell is ever in your way moving forward, my suggestion is that we should just guard them with a Kconfig now and deprecate them in the future. We can demonstrate the features of hawkBit by improving the hawkBit sample.

Would deprecating them unblock #70787?

File: include/zephyr/mgmt/hawkbit/autohandler.h
Comment (maass-hamburg) [reply include/zephyr/mgmt/hawkbit/autohandler.h:32]: the hawkbit api in zephyr is currently not marked as stable (https://docs.zephyrproject.org/latest/develop/api/overview.html), which means we don't have to follow the normal deprecation process.

File: include/zephyr/mgmt/hawkbit/config.h
Comment (maass-hamburg) [reply include/zephyr/mgmt/hawkbit/config.h:182]: `hawkbit_get_action_id` is not a new function, only `hawkbit_get_poll_interval` is, so I would like to leave it like it is.

File: include/zephyr/mgmt/hawkbit/config.h
Comment (maass-hamburg) [reply include/zephyr/mgmt/hawkbit/config.h:56]: It is now what is is.

File: include/zephyr/mgmt/hawkbit/autohandler.h
Comment (ycsin) [reply include/zephyr/mgmt/hawkbit/autohandler.h:32]: Interesting, looks like a lot of stuff there could use an update @nashif

**Comment by maass-hamburg** [comment]
rebased to resolve conflicts

**Comment by aescolar** [comment]
Bulk changing the milestone from everything which did not make it to the v3.7.0 freeze deadline
and which is not labeled as a bug or documentation (or does not appear clearly as only containing
a bug fix or a documentation change).
If you disagree about this change please add the milestone again.

Please note that until rc2 only PRs containing bug fixes, docs and tests can be merged.
After that and until release only bug fixes and docs changes.

Exceptions require approval by the release team and a vote by the TSC.
https://docs.zephyrproject.org/latest/project/release_process.html#stabilization-phase

**Comment by maass-hamburg** [comment]
I like to get forward with this PR. 
First of all as a Maintainer I want to say that:
- the autohandler and the hawkBit shell will stay. 
- useful additions to both of them are welcome.

Both the shell and the autohandler are a great at making hawkBit easy to use and accessible. 
Not everybody want to write their own autohandler and therefore use the autohandler, which will just work and you don't have to do much yourself. And fore the ones, who don't want to use this autohandler, they can, if this PR gets merged, deactivate the included autohandler. 

That being said, I hope we can get this PR approved and merged soon, so we can finally be able to get to https://github.com/zephyrproject-rtos/zephyr/pull/70787

**Comment by maass-hamburg** [comment]
rebased to resolve conflicts

**Comment by maass-hamburg** [comment]
@ycsin are you coming back for a review?
In the dev-review meeting in the last week there where no objections against my changes.

**Comment by maass-hamburg** [comment]
rebased due to unrelated fails in the ci

File: subsys/mgmt/hawkbit/Kconfig
Comment (ycsin) [inline subsys/mgmt/hawkbit/Kconfig:46]: > And fore the ones, who don't want to use this autohandler, they can, if this PR gets merged, deactivate the included autohandler.

Does this dependency means that for user who doesn't want to use the bundled autohandler, they lose access to shell for updating the configurations as well?

**Review by ycsin — COMMENTED** [review]
> @ycsin are you coming back for a review?

Had another look but there seems to be nothing new to review, I'd say my previous review is still up-to-date

> First of all as a Maintainer I want to say that:
> 
> * the autohandler and the hawkBit shell will stay.
> * useful additions to both of them are welcome.
> 
> Both the shell and the autohandler are a great at making hawkBit easy to use and accessible. Not everybody want to write their own autohandler and therefore use the autohandler, which will just work and you don't have to do much yourself. And fore the ones, who don't want to use this autohandler, they can, if this PR gets merged, deactivate the included autohandler.

I agree that the bundled autohandler in this PR might be useful, I have no qualms if this PR updated the one in the samples to make it more portable, which IMHO serves pretty much the same purpose to show the users how to build one, instead of forcing an implementation on them.

I've had my custom implementation of the autohandler in the past, had I chose to upstream and developed it around my use case, folks that came after me probably have to either use it my way or go away, I think that would make hawkBit less inviting

File: subsys/mgmt/hawkbit/Kconfig
Comment (maass-hamburg) [reply subsys/mgmt/hawkbit/Kconfig:46]: Well yes at least when they don't activate the Kconfig for the autohandler. They could still choose to include the autohandler and don't use it theirself.

File: include/zephyr/mgmt/hawkbit/config.h
Comment (maass-hamburg) [reply include/zephyr/mgmt/hawkbit/config.h:186]: this requested change is dismissed, reason above.

**Comment by maass-hamburg** [comment]
> > @ycsin are you coming back for a review?
> 
> Had another look but there seems to be nothing new to review, I'd say my previous review is still up-to-date
> 
> > First of all as a Maintainer I want to say that:
> > 
> > * the autohandler and the hawkBit shell will stay.
> > * useful additions to both of them are welcome.
> > 
> > Both the shell and the autohandler are a great at making hawkBit easy to use and accessible. Not everybody want to write their own autohandler and therefore use the autohandler, which will just work and you don't have to do much yourself. And fore the ones, who don't want to use this autohandler, they can, if this PR gets merged, deactivate the included autohandler.
> 
> I agree that the bundled autohandler in this PR might be useful, I have no qualms if this PR updated the one in the samples to make it more portable, which IMHO serves pretty much the same purpose to show the users how to build one, instead of forcing an implementation on them.
> 
> I've had my custom implementation of the autohandler in the past, had I chose to upstream and developed it around my use case, folks that came after me probably have to either use it my way or go away, I think that would make hawkBit less inviting

@ycsin As I said before the autohandler and the shell are going to stay and if there are good reasons, like here, they are also going to be extended. Requested changes in the other way are hereby dismissed. You can follow the [escalation process](https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-technical-escalation) if you are not okay with that, otherwise I expect a review approving this or requesting changes that are not on this fundamental level.

**Comment by maass-hamburg** [comment]
@jukkar @rlubos can you maybe take a look?

**Review by jukkar — APPROVED** [review]
I am not very familiar with hawkbit so my review is just looking the overall code structure. Anyway, the changes look reasonable.

**Comment by maass-hamburg** [comment]
@kartben there are still requested changes from you, you might want to approve?

**Review by rlubos — APPROVED** [review]

**Review by kartben — APPROVED** [review]

**Pull request merged**
